---
title: "02b_prep_impute_VCF_wPSY2KASP"
output: html_document
date: "2024-02-08"
---
# About
Imputation of PSY2 causal SNP, S1_24155522

-   Inputs: Dart genotypes from HarvestPlus population & additional accessions from CASS project/Cassavabase download (`data/MergedYellowCASSGenotypes.vcf.gz`) and the reference imputation panel (`chr*_RefPanelAndGSprogeny_ReadyForGP_72719.vcf.gz`, available at https://www.cassavabase.org/ftp/marnin_datasets/nextgenImputation2019/ImputationStageIII_72619/)
-   Outputs: Imputered and filtered SNP set **including S1_24155522**, the PSY2 causal SNP, `data/genomat_plusKASP_filtered.RDS`

Based on Marnin Wolfe's [NextGen Cassava GS pipeline](https://github.com/wolfemd/IITA_2021GS/) using tools from the `genomicMateSelectR` package.
(Thanks Marnin!)
Install with: devtools::install_github("wolfemd/genomicMateSelectR", ref = 'master') 

# Setup
```{r}
library(genomicMateSelectR); library(tidyverse); library(magrittr)
library(ggplot2); library(ggrepel); library(data.table)

inpath <- "/workdir/ssv42/mvGWAS/data/"
outName <- "/workdir/ssv42/mvGWAS/data/"
```

If running in Rstudio: set the PATH so we can access bcftools and other software from Rstudio
(Note for future: better solution is to append the programs needed to the path string rather than copying all)
```{r}
# system(paste0("which bcftools"))
# check PATH via: Sys.getenv()
# check what it should be: echo $PATH in terminal
# change PATH via: Sys.setenv(PATH = "/normal/path/to/user/programs/bin/when/in/terminal/")
```


Split the genome-wide VCF to per-chromosome VCF
```{r, eval=F}
require(furrr); plan(multisession, workers = 8)
options(future.globals.maxSize=+Inf); options(future.rng.onMisuse="ignore")
  
vcfIn <- "/workdir/ssv42/mvGWAS/data/MergedYellowCASSGenotypes.vcf.gz"
filters <-"--minDP 4 --maxDP 50" # because using GT not PL for impute (Beagle5)
outPath <- "/workdir/ssv42/mvGWAS/data/"
outSuffix <- "MergedYellowCASSGenotypes"

future_map(1:18,
           ~genomicMateSelectR::splitVCFbyChr(Chr=.,
                                              vcfIn=vcfIn,filters=filters,
                                              outPath=outPath,
                                              outSuffix=outSuffix))
plan(sequential)
```


There are missing sites in the reference panel from adding the KASP markers. First let's try to impute those.
```{r}
library(genomicMateSelectR)
set_wd("/workdir/ssv42/mvGWAS/data/")
refVCF <- "/workdir/ssv42/mvGWAS/data/ImputationRefPanel/chr1_RefPanelAndGSprogeny_ReadyForGP_72719_plusKASP_dedup_sorted.vcf.gz"
mapFile <-  "/workdir/ssv42/mvGWAS/data/ImputationRefPanel/CassavaGeneticMap/chr1_cassava_cM_pred.v6_91019.map"
outName <- "/workdir/ssv42/mvGWAS/data/ImputationRefPanel/chr1_RefPanelAndGSprogeny_ReadyForGP_72719_plusKASP_dedup_sorted_imputed"

nthreads=20;maxmem="500g";impute=TRUE;ne=100000;samplesToExclude=NULL
  system(paste0("java -Xms2g -Xmx",maxmem," -jar /programs/beagle/beagle.jar ",
                "gt=",refVCF," ",
                "map=",mapFile," ",
                "out=",outName," ",
                "nthreads=",nthreads," impute=",impute," ne=",ne,
                ifelse(!is.null(samplesToExclude),paste0(" excludesamples=",samplesToExclude),""))) 
  
  
  
#using version with only the complete set -- just phasing no imputation needed
library(genomicMateSelectR)
set_wd("/workdir/ssv42/mvGWAS/data/")
#refVCF <- "/workdir/ssv42/mvGWAS/data/chr1_RefPanelAndGSprogeny_ReadyForGP_72719_plusKASP_intersectsubset_sort.vcf.gz"
refVCF <- "/workdir/ssv42/mvGWAS/data/chr1_RefPanelAndGSprogeny_ReadyForGP_72719_plusKASP_completesetonly_sorted.vcf.gz"
mapFile <-  "/workdir/ssv42/mvGWAS/data/ImputationRefPanel/CassavaGeneticMap/chr1_cassava_cM_pred.v6_91019.map"
outName <- "/workdir/ssv42/mvGWAS/data/ImputationRefPanel/chr1_RefPanelAndGSprogeny_ReadyForGP_72719_plusKASP_intersectsubset_sorted_phased"

nthreads=20;maxmem="500g";impute=FALSE;ne=100000;samplesToExclude=NULL
  system(paste0("java -Xms2g -Xmx",maxmem," -jar /programs/beagle/beagle.jar ",
                "gt=",refVCF," ",
                "map=",mapFile," ",
                "out=",outName," ",
                "nthreads=",nthreads," impute=",impute," ne=",ne,
                ifelse(!is.null(samplesToExclude),paste0(" excludesamples=",samplesToExclude),""))) 
  
```


First need to append tag ("_A") to sample names so there are not duplicates between reference panel and target VCF apparently (Beagle complained). Used `bcftools reheader` for that.

```{r}
system(paste0("cd /workdir/ssv42/; bcftools query -l chr1_MergedYellowCASS_preimpute_Genotypes.vcf.gz > MergedYellowCASSsamplenames.txt"))

samplenames <- read.table("/workdir/ssv42/MergedYellowCASSsamplenames.txt")
samplenames_tagappend <- paste0(samplenames$V1, "_A")
write.table(samplenames_tagappend, file="/workdir/ssv42/samplenames_tagappend.txt", col.names=F, row.names=F, quote=F)

system(paste0("bcftools reheader -s samplenames_tagappend.txt -o chr1_MergedYellowCASS_preimpute_Genotypes_tagappend.vcf.gz chr1_MergedYellowCASS_preimpute_Genotypes.vcf.gz"))

```



Run imputation for chromosome one
```{r, eval=F}
library(genomicMateSelectR)

targetVCFpath <- "/workdir/ssv42/mvGWAS/data/" # location of the targetVCF
refVCFpath <- "/workdir/ssv42/mvGWAS/data/"
mapPath <-  "/workdir/ssv42/mvGWAS/data/ImputationRefPanel/CassavaGeneticMap/"
outPath <- "/workdir/ssv42/mvGWAS/data/"
outSuffix <- "/workdir/ssv42/mvGWAS/data/MergedYellowCASSImputed"


# genomicMateSelectR::runBeagle5(targetVCF=paste0(targetVCFpath,
#                                 "chr",1, "_MergedYellowCASSGenotypes_tagappend.vcf.gz"),
#                                            refVCF=paste0(refVCFpath,"chr",1,
#                               "_RefPanelAndGSprogeny_ReadyForGP_72719_plusKASP_dedup_sorted_imputed.vcf.gz"),
#                                            mapFile=paste0(mapPath,"chr",1,
#                                                           "_cassava_cM_pred.v6_91019.map"),
#                                            outName=paste0(outPath,"chr",1,
#                                                           "_YellowCASSGenotypes_REF19andKASPimputed"), nthreads=20)



#updated version with only the complete set of overlapped Dart & KASP accessions as the reference (higher DR2 of 0.74 compared to 0.65 for S1_24155522)
refVCFpath <- "/workdir/ssv42/mvGWAS/data/ImputationRefPanel/"
outPath <- "/workdir/ssv42/mvGWAS/data/"

genomicMateSelectR::runBeagle5(targetVCF=paste0(targetVCFpath,
                                "chr",1, "_YellowGenotypes.vcf.gz"), #"_MergedYellowCASSGenotypes.vcf.gz"),
                                           refVCF=paste0(refVCFpath,"chr",1,
"_RefPanelAndGSprogeny_ReadyForGP_72719_plusKASP_completesetonly_sorted_phased.vcf.gz"),
                                           mapFile=paste0(mapPath,"chr",1,
                                                          "_cassava_cM_pred.v6_91019.map"),
                                           outName=paste0(outPath,"chr",1,
                                                          "_YellowGenotypes_REF19andKASPimputed_wcompletesetonly"), nthreads=60)

```

Convert VCF to dosage matrix
```{r}
pathIn <- "/workdir/ssv42/mvGWAS/data/" # location of the targetVCF
pathOut <- "/workdir/ssv42/mvGWAS/data/"

genomicMateSelectR::convertVCFtoDosage(pathIn, pathOut, vcfName="chr1_YellowCASSGenotypes_REF19andKASPimputed_wcompletesetonly")
```
Output: `/workdir/ssv42/mvGWAS/data/chr1_YellowCASSGenotypes_REF19andKASPimputed_wcompletesetonly.raw`



Merge the previous genotype matrix with this one including PSY2 SNP:
```{r}
project_path <- "/workdir/ssv42/mvGWAS/"

genomat_SNPs <- read.table(file=paste0(project_path, "data/GenotypeMatrix_Named_CASS.bimbam"))
colnames(genomat_SNPs) <-  gsub("\\.", "-", colnames(genomat_SNPs))

KASP_imputed_SNPs <- read.table(file=paste0(project_path, "data/chr1_YellowCASSGenotypes_REF19andKASPimputed_wcompletesetonly.raw"), header = T)


#remove genotyping tags
rownames(KASP_imputed_SNPs) <- make.unique(gsub("_A.*", "", KASP_imputed_SNPs$IID) %>% gsub("\\.", "-",.))
#quick fix for missing leading zero in accession name
rownames(KASP_imputed_SNPs)[which(rownames(KASP_imputed_SNPs)=="IITA-TMS-IBA70593")] <- "IITA-TMS-IBA070593"


#remove metadata
KASP_imputed_SNPs_mat <- KASP_imputed_SNPs[,-c(1:6)]


# Extract SNP information from the VCF
system(paste0("bcftools query -f '%ID %ALT  %REF\n' ", project_path, "data/", "chr1_YellowCASSGenotypes_REF19andKASPimputed_wcompletesetonly.vcf.gz > ",
              project_path, "data/KASPimputedSNPinfo.txt"))

# Make a SNP annotation file
system(paste0("bcftools query -f '%ID, %POS, %CHROM\n' ", project_path, "data/", "chr1_YellowCASSGenotypes_REF19andKASPimputed_wcompletesetonly.vcf.gz > ",
              project_path, "data/KASPimputed_SNPannotation.txt"))

# Transpose and join back with SNP info
SNPinfo <- read.table(file=paste0(project_path, "data/KASPimputedSNPinfo.txt"))
colnames(SNPinfo) <- c("SNP", "ALT", "REF")

# Convert to bimbam format
KASP_bimbam <- cbind( SNPinfo, t(KASP_imputed_SNPs_mat))

# Join with original dataframe
genomat_plusKASP <- full_join(genomat_SNPs, (KASP_bimbam %>% filter(SNP == "S1_24155522")))

# Filter to original list of SNPs plus S1_24155522 
#genomat_plusKASP_SNPfilt <- genomat_plusKASP %>% filter(SNP %in% c(genomat_SNPs$SNP, "S1_24155522"))

# Filter to original list of accessions
genomat_plusKASP_filtered <- genomat_plusKASP %>% dplyr::select(all_of(colnames(genomat_SNPs)))

# Add column for compute_group
nSNPs <- dim(genomat_plusKASP_filtered)[1]
ncores <- 80
genomat_plusKASP_filtered$compute_group <- as.numeric(cut_number(c(1:nSNPs), ncores))

# Save new genotype matrix:
saveRDS(genomat_plusKASP_filtered, file="/workdir/ssv42/mvGWAS/data/genomat_plusKASP_filtered.RDS")
```

**NOTE: in this file the favorable allele "A" is the reference allele at S1_24155522 and "C" is the alternate allele, so the effect direction is backwards, i.e. the "alt" allele should have a negative effect on TC and positive on DM**

