---
title: "02c_genotypic_data_curation"
output: html_document
date: "2024-02-07"
---

# About
mvGWAS project
Genotypic data curation. Convert vcf to dosage matrix, examine population structure and LD

-   Inputs: 
        - Imputed genotype data `xx.vcf.gz`
-   Outputs:
        - MAF-filtered dosage matrix `data/DosageMatrix_CASS.RDS`
        - Genotypic principal components `output/DosageMatrix_PCA_CASS_PCmatrix.RDS`
        
# Setup
```{r, echo=F, message=F}
library(genomicMateSelectR); library(slurmR); library(tidyverse); library(magrittr)
library(ggplot2); library(ggrepel); library(data.table)

project_path <- "/workdir/ssv42/mvGWAS/"
setwd(project_path)
set.seed(14850)

Sys.setenv(PATH = "/usr/share/Modules/bin:/programs/docker/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/usr/Arcconf:/programs/bin:/programs/bin/mummer:/programs/bin/util:/programs/bin/bowtie:/programs/bin/bwa:/programs/bin/cufflinks:/programs/bin/tophat:/programs/bin/fastx:/programs/bin/blast:/programs/bin/blat:/programs/bin/perlscripts:/programs/bin/labutils:/programs/bin/gnuplot:/programs/bin/seqclean:/programs/bin/blast+:/programs/bin/sra:/programs/bin/bedtools/bin:/programs/bin/plink:/programs/bin/fastqc:/programs/bin/bowtie2:/programs/bin/clustalw:/programs/bin/rsem:/programs/bin/vcftools:/programs/RepeatMasker:/programs/bin/exonerate/bin:/programs/augustus/bin:/programs/bin/structure:/programs/bin/irods:/programs/bin/bedops:/programs/iv/x86_64/bin:/usr/lib64/openmpi/bin:/programs/texlive/bin/x86_64-linux:/programs/R-4.0.5-r9/bin:/programs/samtools-1.18/bin:/programs/bcftools-1.18/bin:/programs/htslib-1.18/bin:/home/ssv42/.local/bin:/home/ssv42/bin:/usr/Arcconf")
```


# Genotypic data curation
Select which dataset to work with:
```{r}
#Select dataset
dataset <- "CASS" #"HP"
print(paste0("working with ", dataset, " data"))


if(dataset=="CASS"){
  VCFname <- "AllChrom_MergedYellowCASSGenotypes_REF19imputedFiltered_deduped"
} else if(dataset=="HP"){
  VCFname <- "AllChrom_YellowGenotypes_REF19imputedAndFiltered_noITH"
}
```


Convert to dosage matrix:
```{r}
genomicMateSelectR::convertVCFtoDosage(pathIn = paste0(project_path, "data/"), 
                                       vcfName = VCFname, 
                                       pathOut = paste0(project_path, "data/"))

#read in dosage matrix 
dosages <- read.table( paste0( project_path, "/data/", VCFname, ".raw"), header=T)

#remove genotyping tag
rownames(dosages) <- make.unique(gsub("_A.*", "", dosages$IID) %>% gsub("\\.", "-", .))
#remove metadata
dosagematrix <- dosages[,-c(1:6)]

#Round everything to {0, 1, 2}
dosagematrix_rounded <- round(dosagematrix)

#Old code before MAF filter was applied at post-imputation filtering step: save raw dosage matrix
#saveRDS(dosagematrix_rounded, file= paste0(project_path, "data/DosageMatrix_preMAFfiltering", dataset, ".RDS")) 



#confirm that there are no NAs in the matrix 
NAvalues <- sapply(dosagematrix_rounded, is.na, simplify="array")
length(which(NAvalues)) == 0

#filter for MAF > 0.05
MAF <- sapply(dosagematrix_rounded, sum) / (nrow(dosagematrix_rounded)*2) #calculate MAF
MAF_tokeep <- MAF > 0.05 & MAF < 0.95
MAFdf <- as.data.frame(MAF) %>% mutate("SNP" = gsub("_[A-Z]$", "", rownames(.)))
saveRDS(MAFdf, file=paste0(project_path, "data/minor_allele_frequencies.RDS"))

dosagematrix_MAFfilt <- dosagematrix_rounded[,MAF_tokeep]

saveRDS(dosagematrix_MAFfilt, file= paste0(project_path, "data/DosageMatrix_", dataset, ".RDS")) 

write.table(rownames(dosagematrix_MAFfilt), col.names=F, row.names=F, quote=F, 
            file= paste0(project_path, "data/", dataset, "_accessionnames.txt"))

write.table(colnames(dosagematrix_MAFfilt), col.names=F, row.names=F, quote=F, 
            file= paste0(project_path, "data/", dataset, "_SNPorder_MAF05.txt"))
```



Extract SNP information from the VCF
```{r}
# Extract SNP information from the VCF
system(paste0("bcftools query -f '%ID %ALT  %REF\n' ", project_path, "data/", VCFname, ".vcf.gz > ",
              project_path, "data/SNPinfo_", dataset, ".txt"))


# Make a SNP annotation file
system(paste0("bcftools query -f '%ID, %POS, %CHROM\n' ", project_path, "data/", VCFname, ".vcf.gz > ",
              project_path, "data/SNPannotation_", dataset, ".txt"))

```



Calculate PCA
```{r}
#principal component analysis
dosagesPCA <- prcomp(dosagematrix_MAFfilt)
saveRDS(dosagesPCA, paste0(project_path, "data/DosageMatrix_PCA_", dataset, ".RDS")) 
```


Examine population structure
```{r}
dosages <- readRDS(file= paste0(project_path, "data/DosageMatrix_", dataset, ".RDS"))
accessionnames <- read.table(file= paste0(project_path, "data/", dataset, "_accessionnames.txt"))[,1]

dosagesPCA <- readRDS(file=paste0(project_path, "data/DosageMatrix_PCA_", dataset, ".RDS"))
dosagesPC_sum <- summary(dosagesPCA) 

dosagesPC <- as.data.frame(dosagesPCA$x); 
dosagesPC$germplasmName <- accessionnames

if(dataset == "CASS"){
  dosagesPC$population <- c(rep("CB",50), rep("HP",nrow(dosagesPC)-50))
} else if(dataset=="HP"){
  dosagesPC$population <- c(rep("HP",nrow(dosagesPC)))
}


ggplot(data=dosagesPC, aes(PC1, PC2)) +
  geom_point(size=0.8, nudge_x = 5, aes(color=population)) +
  xlab(paste0("PC1, ", (dosagesPC_sum$importance["Proportion of Variance",1])*100, 
              "% variance explained")) +
  ylab(paste0("PC2, ", (dosagesPC_sum$importance["Proportion of Variance",2])*100, 
              "% variance explained"))
  

#Are these separated by Latin American ancestry? Yes, PC1 is generally Latin American ancestry
pedigrees <- read.table("./data/pedigrees.txt", sep="\t", header=T) %>%
  #mutate(Accession = gsub("-","\\.", Accession)) %>%
  mutate(LaAm = as.factor(case_when(grepl("HIL", Female_Parent) == TRUE & 
                                      grepl("HIL", Male_Parent) == TRUE~ "1",
                          grepl("HIL", Female_Parent) == TRUE & Cross_Type == "open" ~ "1",          
                          grepl("IITA", Female_Parent) & grepl("CM", Male_Parent) == TRUE ~ "1",
                          grepl("IITA", Female_Parent) & grepl("GM", Male_Parent) == TRUE ~ "1",
                          grepl("BRANCA", Male_Parent) == TRUE ~ "1",
                          grepl("SM", Female_Parent) == TRUE & Male_Parent == "" ~ "1",
                          grepl("MTAI-8", Female_Parent) == TRUE | grepl("MTAI-8", Male_Parent) == TRUE~ "1",
                          grepl("SM", Female_Parent) == TRUE & grepl("C-", Male_Parent) == TRUE| 
                             grepl("GM", Female_Parent) == TRUE | grepl("GM", Male_Parent) == TRUE|
                            grepl("CM", Female_Parent) == TRUE | grepl("CM", Male_Parent) == TRUE|
                            grepl("C-", Female_Parent) == TRUE | grepl("C-", Male_Parent) == TRUE|
                            grepl("COL", Female_Parent) == TRUE | grepl("COL", Male_Parent) == TRUE|
                            grepl("Abbey", Female_Parent) == TRUE ~ "2",
                          is.na(Female_Parent) & is.na(Male_Parent) ~ "NA",
                          Female_Parent == "" & Male_Parent == "" ~ "NA",
                          TRUE ~ "0")))

PCAdf_ped <- dosagesPC %>% dplyr::select(c(1:10, ncol(dosagesPC)-1,ncol(dosagesPC))) %>% 
  arrange(PC1) %>% 
  left_join(pedigrees, by=c("germplasmName"="Accession"))


ggplot(data=PCAdf_ped, aes(PC1, PC2)) +
  geom_point(aes(color=LaAm), size=0.8) +
  xlab(paste0("PC1, ", round(dosagesPC_sum$importance["Proportion of Variance",1],2)*100, 
              "% variance explained")) +
  ylab(paste0("PC2, ", round(dosagesPC_sum$importance["Proportion of Variance",2],2)*100, 
              "% variance explained")) +
  geom_text(aes(label=germplasmName), size=1, nudge_x=7.5) +
  scale_color_manual(values=c("#FA4C4C", "#BE5DFF",  "#5D80FF", "#8D8D8D"), 
                     labels=c("0", "1", "2", "unknown"),
                     name="Number of L.Am. parents")

ggsave(filename=paste0(project_path, "figures/PCAplot_ancestry_", dataset, ".jpg"))


table(PCAdf_ped$LaAm)
```

#Look at PC loadings
```{r}
accessionnames <- read.table(file= paste0(project_path, "data/", dataset, "_accessionnames.txt"))[,1]
dosagesPCA <- readRDS(file=paste0(project_path, "data/DosageMatrix_PCA_", dataset, ".RDS"))
dosagesPC_sum <- summary(dosagesPCA) 

dosagesPC <- as.data.frame(dosagesPCA$x); dosagesPC$germplasmName <- accessionnames

Vexplained <- dosagesPC_sum$importance

loadings <- as.data.frame(dosagesPCA$rotation[,1:10])

#confirm correct dimensions and reattach SNP names:
SNPorder <- read.table(paste0("./data/", dataset, "_SNPorder_MAF05.txt"))
dim(loadings)[1] == dim(SNPorder)[1]
rownames(loadings) <- SNPorder[,1]
loadings$SNP_ID <- SNPorder[,1]

#sort by largest absolute value
PC1loadings_sort <- arrange(loadings, abs(PC1))
PC2loadings_sort <- arrange(loadings, abs(PC2))
PC3loadings_sort <- arrange(loadings, abs(PC3))
PC4loadings_sort <- arrange(loadings, abs(PC4))
PC5loadings_sort <- arrange(loadings, abs(PC5))

toploadingSNPs <- c(PC1loadings_sort$SNP_ID[1:20], PC2loadings_sort$SNP_ID[1:20])
toploadingSNPsPCs <- loadings %>% filter(SNP_ID %in% toploadingSNPs)

#View(toploadingSNPsPCs)
#I don't think PSY2 SNP is here. Only S1_S1_19510220 shows up as a top loading.
#There are lots of SNPs on chr 5 and 13 with high PC loadings.

#PC loadings by chromosome:
loadings_pos <- loadings %>% mutate(POS = gsub("S[0-9]_", "", SNP_ID),
                    CHR =  gsub("_.*", "", SNP_ID) %>% gsub("S","",.)) %>%
  #pivot_longer(cols=c(PC1, PC2, PC3, PC4, PC5, PC6, PC7, PC8, PC9, PC10)) %>%
  arrange(CHR, POS)

#set positions for axis ticks centered by chromosome
xaxisticks <- loadings_pos %>% group_by(CHR) %>% summarise(midpt = nth(SNP_ID, round(n()/2))) %>% as.data.frame() 

#set color scheme
colPalette <- rep(c("blue4","pink3"),9)

for(i in c(1:10)){
  
  PCi <- paste0("PC",i)
  
  ## Plot PCs by chromosome
  loadingsplot <- ggplot(loadings_pos, aes(x=SNP_ID, y=get(PCi), color=CHR)) +
    geom_point(size=0.5) +
    labs(y=paste0(PCi," loading")) +
    scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
    ggtitle(paste0(PCi," loadings, ", Vexplained[2,i]*100, "% variance explained")) +
    scale_color_manual(values=colPalette) +
    guides(color="none") +
    geom_vline(xintercept="S1_24155945", color="orange", lty="dashed") #closest to PSY2
  
  print(loadingsplot)
  
  ggsave(filename=paste0("./figures/", PCi, "_loadings_", dataset, "_", Sys.Date(),".jpg"))
}

#PC 3 seems to capture the chr 1 introgression
#note: Latin American ancestry plot is in 02_genotypic_data_curation.Rmd
```




Examine LD structure on chromosome 1:
```{r, eval=F}
# dosagematrix_rounded <- readRDS(file= paste0(project_path, "data/DosageMatrix_", dataset, ".RDS")) 
# 
# # Subset for chromosome 1
# chr1_dosage <- dosagematrix_rounded %>% dplyr::select_if(grepl("S1_.*", colnames(dosagematrix_rounded)))
# 
# # Calculate LD by correlation coefficient between markers
# chr1_LD <- cor(chr1_dosage)
# saveRDS(chr1_LD, file=paste0(project_path, "data/Chr1_LD_", dataset, ".RDS"))
chr1_LD <- readRDS(file=paste0(project_path, "data/Chr1_LD_", dataset, ".RDS"))

# 1_24155945 is the marker closest to PSY2 which is marker 2492 on chr 1
png(filename=paste0(project_path, "figures/chr1LDheatmap_", dataset, "_", format(Sys.time(), "%Y%m%d_%H%M"), ".png"),
    width=4000, height=4000, units="px")
heatmap(chr1_LD, Rowv=NA, Colv=NA)
dev.off()


## ggplot method
chr1_LD_melted <- reshape2::melt(chr1_LD) 

ggplot(chr1_LD_melted, aes(Var1, Var2, fill=value)) +
  geom_tile(lwd=0) +
  theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank()) +
  scale_fill_gradient(low = "navyblue", high = "red")

ggsave(filename=paste0(project_path, "figures/chr1LD_GGheatmap_", dataset, "_", format(Sys.time(), "%Y%m%d_%H%M"), ".jpg"))

```


Examine LD decay
```{r}
dosagematrix_rounded <- readRDS(file= paste0(project_path, "data/DosageMatrix_", dataset, ".RDS")) 

LDallchrom <- data.frame(var1=NA, var2=NA, cor=NA, chrom=NA)
#split by chromosome so it's easier to compute
for(i in 1:18){
  print(i)
  # Subset for chromosome i
  chri_dosage <- dosagematrix_rounded %>% dplyr::select_if(
    grepl(paste0("S",i,"_.*"), colnames(dosagematrix_rounded)))
  
  chri_LD <- cor(chri_dosage)
  
  chri_LD_long <- rstatix::cor_gather(chri_LD) %>%
    dplyr::select(var1, var2, cor) %>%
    mutate(chrom = i)
  
  LDallchrom <- rbind(LDallchrom, chri_LD_long)
  saveRDS(LDdecay, file=paste0(project_path, "output/LD_stats.RDS"))
}


```



PCA on haplotype matrix
```{r}
#read in haplotypes
haps <- readRDS(file=paste0(project_path, "data/haplotype_matrix_", dataset, ".rds"))

#subset for chromosome 1
haps_chr1 <- haps[,grepl("^1_", colnames(haps))]

#SNP names
chr1_SNP_pos <- as.numeric(gsub("^1_", "", colnames(haps_chr1)) %>% gsub("_._.","",.))

#subset for PSY2 region: (PSY2 v6 is 24153419 - 24156720, add cushion of 20kb on either side)
PSY2region <- (chr1_SNP_pos > 24133419 & chr1_SNP_pos < 24176720)
#PSY2regionIDMs <- c(24150027, 24154529, 24155085, 24155561, 24156365)
haps_PSY2region <- haps_chr1[, PSY2region]
saveRDS(haps_PSY2region, file=paste0(project_path, "data/Haplotypes_PSY2region_", dataset, ".RDS"))

#'what if' larger PSY2 region: +/- 100kb  24053419 to 24256720
whatif_region <- (chr1_SNP_pos > 24053419 & chr1_SNP_pos < 24256720)
haps_whatif_region <- haps_chr1[, whatif_region]


#subset for introgression:  ( 24Mb to the end, v6)
IDMlist <- read.csv(paste0(project_path,"data/IDM_list_all_strict_v6.csv"), header=T)
haps_chr1_IDM <- haps_chr1[,chr1_SNP_pos %in% as.numeric(IDMlist$Pos)] %>% as.data.frame()
saveRDS(haps_chr1_IDM, file=paste0(project_path, "data/Haplotypes_IDMs_", dataset, ".RDS"))


#PCA haplotypes
PCA_haps_chr1 <- prcomp(haps_chr1)
saveRDS(PCA_haps_chr1, paste0(project_path, "data/Chr1Haplotype_PCA_", dataset, ".RDS")) 

```

Cluster haplotypes:
```{r}
#How many clusters capture the variation in chr. 1 haplotypes?
#factoextra::fviz_nbclust(haps_chr1, FUNcluster=kmeans, k.max=20, method = "wss")
#ggsave(filename=paste0(project_path, "figures/Haplotype_cluster_nbclust_", dataset, ".jpg"))

factoextra::fviz_nbclust(haps_PSY2region, FUNcluster=kmeans, k.max=10, method = "wss")
factoextra::fviz_nbclust(haps_whatif_region, FUNcluster=kmeans, k.max=10, method = "wss")


factoextra::fviz_nbclust(haps_chr1_IDM, FUNcluster=kmeans, k.max=20, method = "wss")

if(dataset == "HP"){ 
  nclusters_chr1 <- 17; nclusters_IDM <- 6; nclusters_PSY2 <- 3}
if(dataset == "CASS") {
  nclusters_chr1 <- 12; nclusters_IDM <- 5; nclusters_PSY2 <- 3}

#Chr1clusters <- kmeans(haps_chr1, centers=nclusters_chr1, iter.max = 10, nstart = 50)
PSY2clusters <- kmeans(haps_PSY2region, centers=nclusters_PSY2, iter.max = 10, nstart = 50)
saveRDS(PSY2clusters, file=paste0(project_path, "output/PSY2haplotype_kmeans_clusters.RDS"))


whatifclusters <- kmeans(haps_whatif_region, centers=4, iter.max = 10, nstart = 50)
saveRDS(whatifclusters, file=paste0(project_path, "output/whatif_PSY2haplotype_kmeans_clusters.RDS"))


IDMclusters <- kmeans(haps_chr1_IDM, centers=nclusters_IDM, iter.max = 10, nstart = 50)
```



Visualize PCA and K-means Clusters:
```{r, eval=T}
PCA_haps_chr1_df <- as.data.frame(PCA_haps_chr1$x) 

Haps_chr1 <- PCA_haps_chr1_df %>%
  mutate(IDName = rownames(PCA_haps_chr1_df),
         haplotypeName = gsub( "_A[0-9]*_Hap", "_Hap", rownames(PCA_haps_chr1_df)),
         germplasmName = gsub( "_A[0-9]*_", "_", rownames(PCA_haps_chr1_df)) %>% 
           gsub( "_Hap." , "", .) %>%
           gsub("\\.", "-", .)) %>%
  mutate(PSY2clusters = as.factor(PSY2clusters[[1]]),
         whatifclusters = as.factor(whatifclusters[[1]]))
         #IDMclusters = as.factor(IDMclusters[[1]]))


ggplot(data=Haps_chr1, aes(PC2, PC3, color=whatifclusters)) +
  geom_point(size=0.8) +
  xlab(paste0("PC2, ", (dosagesPC_sum$importance["Proportion of Variance",1])*100, 
              "% variance explained")) +
  ylab(paste0("PC3, ", (dosagesPC_sum$importance["Proportion of Variance",2])*100, 
              "% variance explained"))


ggplot(data=Haps_chr1, aes(PC2, PC3, color=PSY2clusters)) +
  geom_point(size=0.8) +
  xlab(paste0("PC2, ", (dosagesPC_sum$importance["Proportion of Variance",1])*100, 
              "% variance explained")) +
  ylab(paste0("PC3, ", (dosagesPC_sum$importance["Proportion of Variance",2])*100, 
              "% variance explained"))

#saveRDS(Haps_chr1, file=paste0(project_path, "data/Chr1_haplotype_clusters_", dataset, ".RDS"))

```

Make haplotype cluster incidence matrices for fixed effect covariates:
```{r}
#wide
PSY2haps_incidence <- Haps_chr1 %>% dplyr::select(germplasmName, PSY2clusters) %>%
  reshape2::melt(id.vars="germplasmName") %>%
  reshape2::dcast(formula = germplasmName ~ value, fun.aggregate = length)
colnames(PSY2haps_incidence) <- c("germplasmName", paste0("PSY2hap_", colnames(PSY2haps_incidence)[-1]))

#long
PSY2haps_incidence_long <- Haps_chr1 %>% dplyr::select(germplasmName, IDName, PSY2clusters) %>%
  mutate(Hap = gsub(".*_Hap","Hap", IDName)) %>%
  pivot_wider(id_cols=germplasmName, names_from=Hap, values_from=PSY2clusters) %>%
  mutate(HapA = as.factor(HapA), HapB = as.factor(HapB))

saveRDS(PSY2haps_incidence_long, file=paste0(project_path, "output/PSY2haplotype_3cluster_levels.RDS"))


whatifPSY2haps_incidence_long <- Haps_chr1 %>% dplyr::select(germplasmName, IDName, whatifclusters) %>%
  mutate(Hap = gsub(".*_Hap","Hap", IDName)) %>%
  pivot_wider(id_cols=germplasmName, names_from=Hap, values_from=whatifclusters) %>%
  mutate(HapA = as.factor(HapA), HapB = as.factor(HapB))

saveRDS(whatifPSY2haps_incidence_long, file=paste0(project_path, "output/whatifPSY2haplotype_100kb_4cluster_levels.RDS"))





IDMhaps_incidence <- Haps_chr1 %>% dplyr::select(germplasmName, IDMclusters) %>%
  reshape2::melt(id.vars="germplasmName") %>%
  reshape2::dcast(formula = germplasmName ~ value, fun.aggregate = length)
colnames(IDMhaps_incidence) <- c("germplasmName", paste0("IDMhap_", colnames(IDMhaps_incidence)[-1]))

Chr1HapCovMat <- left_join(PSY2haps_incidence, IDMhaps_incidence, by="germplasmName")
saveRDS(Chr1HapCovMat, file=paste0(project_path, "data/Chr1_hap_incidence_mat", dataset, ".RDS"))
```



Make full haplotype incidence matrices (wide) for random effect covariates:
```{r}
#PSY2 region
haps_PSY2region_df <- as.data.frame(haps_PSY2region) %>% 
  mutate(germplasmName = gsub( "_A[0-9]*_", "_", rownames(haps_PSY2region)) %>%
         gsub( "_Hap." , "", .) %>% gsub("\\.", "-", .))

PSY2haps_factors <- as.data.frame(haps_PSY2region) %>% unique() %>%
  mutate(HAP_ID = as.factor(1:nrow(unique(haps_PSY2region)))) 

PSY2haps_incidence <- left_join(haps_PSY2region_df, PSY2haps_factors) %>% #joining by all matching columns
  dplyr::select(germplasmName, HAP_ID) %>%
  reshape2::melt(id.vars="germplasmName") %>%
  reshape2::dcast(formula = germplasmName ~ value, fun.aggregate = length)
colnames(PSY2haps_incidence) <- c("germplasmName", paste0("PSY2hap_", colnames(PSY2haps_incidence)[-1]))
rownames(PSY2haps_incidence) <- PSY2haps_incidence$germplasmName

saveRDS(PSY2haps_incidence, file=paste0(project_path, "data/PSY2_hap_incidence_mat", dataset, ".RDS"))



#Introgression region

haps_IDMs_df <- as.data.frame(haps_chr1_IDM) %>% 
  mutate(germplasmName = gsub( "_A[0-9]*_", "_", rownames(haps_chr1_IDM)) %>%
         gsub( "_Hap." , "", .) %>% gsub("\\.", "-", .))

IDMhaps_factors <- as.data.frame(haps_chr1_IDM) %>% unique() %>%
  mutate(HAP_ID = as.factor(1:nrow(unique(haps_chr1_IDM)))) 

IDMhaps_incidence <- left_join(haps_IDMs_df, IDMhaps_factors) %>% #joining by all matching columns
  dplyr::select(germplasmName, HAP_ID) %>%
  reshape2::melt(id.vars="germplasmName") %>%
  reshape2::dcast(formula = germplasmName ~ value, fun.aggregate = length)
colnames(IDMhaps_incidence) <- c("germplasmName", paste0("IDMhap_", colnames(IDMhaps_incidence)[-1]))
rownames(IDMhaps_incidence) <- IDMhaps_incidence$germplasmName

saveRDS(IDMhaps_incidence, file=paste0(project_path, "data/IDM_hap_incidence_mat", dataset, ".RDS"))

```


Make full haplotype incidence matrices (one for each haplotype A and B) for random effect covariates:

```{r}
#PSY2 region haplotypes:
haps_PSY2region_df <- as.data.frame(haps_PSY2region) %>% 
  mutate(haplotypeName = gsub( "_A[0-9]*_", "_", rownames(haps_PSY2region))) %>%
  mutate(germplasmName = gsub("_Hap[A-Z]","", haplotypeName) %>% gsub("\\.","-",.)) %>%
  left_join(PSY2haps_factors)

hapA_PSY2 <- haps_PSY2region_df %>% filter(grepl("HapA", haplotypeName)) %>%
  dplyr::select(germplasmName, HAP_ID) %>%
  rename(PSY2_HAPA = HAP_ID)

hapB_PSY2 <- haps_PSY2region_df %>% filter(grepl("HapB", haplotypeName))  %>%
  dplyr::select(germplasmName, HAP_ID) %>%
  rename(PSY2_HAPB = HAP_ID)

#IDM haplotypes:
haps_IDMs_df <- as.data.frame(haps_chr1_IDM) %>% 
  mutate(haplotypeName = gsub( "_A[0-9]*_", "_", rownames(haps_chr1_IDM))) %>%
  mutate(germplasmName = gsub("_Hap[A-Z]","", haplotypeName) %>% gsub("\\.","-",.)) %>%
  left_join(IDMhaps_factors)

hapA_IDMs <- haps_IDMs_df %>% filter(grepl("HapA", haplotypeName)) %>%
  dplyr::select(germplasmName, HAP_ID) %>%
  rename(IDMs_HAPA = HAP_ID)
hapB_IDMs <- haps_IDMs_df %>% filter(grepl("HapB", haplotypeName))  %>%
  dplyr::select(germplasmName, HAP_ID) %>%
  rename(IDMs_HAPB = HAP_ID)

allHaps <- left_join(hapA_PSY2, hapB_PSY2) %>%
  left_join(hapA_IDMs) %>%
  left_join(hapB_IDMs)

saveRDS(allHaps, file=paste0(project_path, "data/Haplo_Factors_", dataset, ".RDS"))
```




Make haplotype relationship matrices for random effect covariate:
```{r}
rownames(PSY2haps_factors) <- PSY2haps_factors$HAP_ID
Amat_haps_PSY2region <- rrBLUP::A.mat(PSY2haps_factors %>% dplyr::select(-"HAP_ID"))
dim(Amat_haps_PSY2region)
#colnames(Amat_haps_PSY2region) <- paste0("PSY2hap_", colnames(Amat_haps_PSY2region))
#rownames(Amat_haps_PSY2region) <- paste0("PSY2hap_", rownames(Amat_haps_PSY2region))
saveRDS(Amat_haps_PSY2region, file=paste0(project_path, "data/Amat_PSY2haplotypes_", dataset, ".RDS"))

rownames(IDMhaps_factors) <- IDMhaps_factors$HAP_ID
Amat_haps_chr1_IDM <-rrBLUP::A.mat(IDMhaps_factors %>% dplyr::select(-"HAP_ID"))
dim(Amat_haps_chr1_IDM)
#colnames(Amat_haps_chr1_IDM) <- paste0("IDMhap_", colnames(Amat_haps_chr1_IDM))
#rownames(Amat_haps_chr1_IDM) <- paste0("IDMhap_", rownames(Amat_haps_chr1_IDM))
saveRDS(Amat_haps_chr1_IDM, file=paste0(project_path, "data/Amat_IDMhaplotypes_", dataset, ".RDS"))

```



Look at which of the PSY2 clusters seems to capture the causal variant
```{r}
 BLUPs_hap_data <- readRDS(file=paste0(project_path, "data/BLUPs_scaled_and_hap_data_", dataset, ".RDS"))

haps_phenos <- left_join(BLUPs_hap_data, Haps_chr1 , by="germplasmName")

ggplot(haps_phenos, aes(TCICHK, DM, color=PSY2clusters)) +
  geom_point()
```

Cluster PSY2 haplotypes with PCA with both haplotypes together:

```{r}
#read in dosage matrix
dosagematrix_MAFfilt <- readRDS(file= paste0(project_path, "data/DosageMatrix_", dataset, ".RDS"))
colnames(dosagematrix_MAFfilt) <- gsub("_[A-Z]$", "", colnames(dosagematrix_MAFfilt))

#read in PSY2 haplotype region for SNP names
haps_PSY2region <- readRDS(file=paste0(project_path, "data/Haplotypes_PSY2region_", dataset, ".RDS"))
PSY2regionSNPs <- paste0("S", gsub("_[A-Z]_[A-Z]$", "", colnames(haps_PSY2region)))

#PCA with PSY2 region SNPs
#PSY2_dosagemat <- dosagematrix_MAFfilt %>% dplyr::select(any_of(PSY2regionSNPs))
PSY2cols <- range(which(colnames(dosagematrix_MAFfilt) %in% PSY2regionSNPs)) + c(-20, 20)
PSY2_dosagemat <- dosagematrix_MAFfilt[,c(PSY2cols[1]:PSY2cols[2])]

PSY2_dosage_PCA <- prcomp(PSY2_dosagemat)
saveRDS(PSY2_dosage_PCA, file=paste0(project_path, "output/PSY2_dosagematrix_PCA.RDS"))


#Merge with phenotype info
PSY2_dosage_PCA_df <- as.data.frame(PSY2_dosage_PCA$x)[,1:10]
colnames(PSY2_dosage_PCA_df) <- paste0("PSY2_", colnames(PSY2_dosage_PCA_df))
PSY2_dosage_PCA_df$germplasmName <- rownames(PSY2_dosage_PCA_df)

BLUPs_hap_data_PSY2PCs <- left_join(BLUPs_hap_data, PSY2_dosage_PCA_df, by="germplasmName")


ggplot(BLUPs_hap_data_PSY2PCs, aes(TCICHK, DM, color=PSY2_PC1)) +
  geom_point()
```




Subset for ITH accessions:
bcftools view --samples-file /workdir/ssv42/mvGWAS/data/ITHsamplenames.txt /workdir/ssv42/mvGWAS/data/MergedYellow_KASP_Genotypes.vcf.gz > MergedYellow_KASP_Genotypes_ITHonly.vcf.gz



