---
title: "03_mvGEMMA_analysis"
output: html_document
date: "2024-02-08"
---

# About
Run multi-trait associations in GEMMA.
  - Inputs: Imputed and filtered genotype file, `AllChrom_YellowGenotypes_REF19imputedAndFiltered.vcf.gz`; deregressed BLUPs `alltrait_drgBLUPs.RDS`
  - Outputs: List of SNPs and p-values, Manhattan plots, and QQ plots for GEMMA MV associations and univariate associations

# Setup
```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse); library(qqman)
project_path <- "/workdir/ssv42/mvGWAS/"
setwd(project_path)
set.seed(14850)

# Set the PATH so we can access bcftools and other software from Rstudio
Sys.setenv(PATH = "/usr/share/Modules/bin:/programs/docker/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/usr/Arcconf:/programs/bin:/programs/bin/mummer:/programs/bin/util:/programs/bin/bowtie:/programs/bin/bwa:/programs/bin/cufflinks:/programs/bin/tophat:/programs/bin/fastx:/programs/bin/blast:/programs/bin/blat:/programs/bin/perlscripts:/programs/bin/labutils:/programs/bin/gnuplot:/programs/bin/seqclean:/programs/bin/blast+:/programs/bin/sra:/programs/bin/bedtools/bin:/programs/bin/plink:/programs/bin/fastqc:/programs/bin/bowtie2:/programs/bin/clustalw:/programs/bin/rsem:/programs/bin/vcftools:/programs/RepeatMasker:/programs/bin/exonerate/bin:/programs/augustus/bin:/programs/bin/structure:/programs/bin/irods:/programs/bin/bedops:/programs/iv/x86_64/bin:/usr/lib64/openmpi/bin:/programs/texlive/bin/x86_64-linux:/programs/R-4.0.5-r9/bin:/programs/samtools-1.18/bin:/programs/bcftools-1.18/bin:/programs/htslib-1.18/bin:/home/ssv42/.local/bin:/home/ssv42/bin:/usr/Arcconf")
```

Select which dataset to work with:
```{r}
#Select dataset
dataset <- "CASS" #"HP"
TCtrait <- "TCICHK"
filter_trial_byH2 <- "FALSE"
print(paste0("working with ", dataset, " data"))


if(dataset=="CASS"){
  VCFname <- "AllChrom_MergedYellowCASSGenotypes_REF19imputedFiltered_deduped"
} else if(dataset=="HP"){
  VCFname <- "AllChrom_YellowGenotypes_REF19imputedAndFiltered"
}
```



## Subset for accessions with genotype data
```{r}
# Read in the deregressed BLUPs
BLUPs <- readRDS(file=paste0(project_path, "data/alltrait_scaled_drgBLUPs_", dataset, "_", TCtrait, "_", "H2filter", filter_trial_byH2,  ".RDS"))


# Read in names of genotyped samples
accessionnames <- read.table(file= paste0(project_path, "data/", dataset, "_accessionnames.txt"))[,1]

# Subset for BLUPs with genotype data
BLUPs_mvsubset <- BLUPs %>% filter(germplasmName %in% accessionnames)
print(paste0("Subset of ", nrow(BLUPs_mvsubset), " accessions"))
saveRDS(BLUPs_mvsubset, file=paste0(project_path, "data/alltrait_scaled_drgBLUPs_genodsubset_", dataset, ".RDS"))

# Save names
genophenonames <- BLUPs_mvsubset$germplasmName
```



## Convert genotype matrix to bimbam format
```{r}
# Read in genotype dosage matrix
genodata <- readRDS(file=paste0(project_path, "data/DosageMatrix_", dataset, ".RDS"))

# Reattach sample names
nrow(genodata) == length(accessionnames) #confirm correct dimensions
rownames(genodata) <- accessionnames

# Subset
genodata_subset <- genodata %>% filter(rownames(genodata) %in% genophenonames)
nrow(genodata_subset) == length(genophenonames) #confirm proper dimensions

# Read in original SNP info and subset for those that passed MAF threshold
SNPinfo <- read.table(file=paste0(project_path, "data/SNPinfo_", dataset, ".txt"))
colnames(SNPinfo) <- c("SNP", "ALT", "REF")
SNPinfo_MAFfilt <- SNPinfo[(SNPinfo$SNP %in% gsub("_[A-Z]$", "", colnames(genodata))),]
nrow(SNPinfo_MAFfilt) == ncol(genodata_subset) #confirm dimensions

# Convert to bimbam format
bimbam <- cbind( SNPinfo_MAFfilt,
                  t(genodata_subset))
write.table(bimbam, file=paste0(project_path, "data/GenotypeMatrix_", dataset, ".bimbam"),
          col.names = FALSE, row.names=FALSE, quote=FALSE)

# Save named version for Stephens analysis
write.table(bimbam, file=paste0(project_path, "data/GenotypeMatrix_Named_", dataset, ".bimbam"),
          col.names = T, row.names=T, quote=FALSE)

# Save list of sample names
sample_names <- colnames(bimbam)[-c(1:3)]
saveRDS(sample_names, file=paste0(project_path, "data/BimbamSampleNames_",dataset,".RDS"))


# Save list of SNP names
bimbamSNPorder <- bimbam$SNP
write.table(bimbamSNPorder, file=paste0(project_path, "data/SNPorder_", dataset, ".txt", col.names=F, row.names=F)) 

```


## Generate a phenotypic data matrix in bimbam format
Genotype file sample names and phenotype file sample names must be in the same order!!
```{r}
# Match with BLUPs to get them in the right order and subset for genotyped accessions
Ordered_BLUPs <- left_join(as.data.frame(sample_names), 
                           BLUPs, by=c("sample_names" = "germplasmName"))
  
# Check that order and dimensions are still correct
nrow(Ordered_BLUPs) == length(sample_names)
sum(Ordered_BLUPs$germplasmName != sample_names) == 0


# Write out phenotypes in format for GEMMA
write.table(Ordered_BLUPs[,-c(1)],
            file=paste0(project_path, "data/BLUPmatrix_scaled_", dataset, ".txt"), 
            col.names=F, row.names=F)


# Write out phenotypes with names attached for Stephens analysis
write.table(Ordered_BLUPs, 
            file=paste0(project_path, "data/BLUPmatrix_scaled_named_", dataset, ".txt"), 
            col.names=T)

# confirm correct sample names and order
sum(Ordered_BLUPs$sample_names != colnames(bimbam)[-c(1:3)]) == 0
```


## Generate covariates file
```{r}
# Read in PC decomposition
genoPCA <- readRDS(file=paste0(project_path, "data/DosageMatrix_PCA_", dataset, ".RDS"))
genoPCs <- as.data.frame(genoPCA$x)
sample_names <- readRDS(file=paste0(project_path,"data/BimbamSampleNames_",dataset,".RDS"))

# Make covariates matrix with first x PCs
nPCs <- 5
PreCovMat <- as.data.frame(cbind(rep(1,dim(genoPCs)[1]), genoPCs[,1:nPCs]))
colnames(PreCovMat)[1] <- "intercept"
PreCovMat$germplasmName <- rownames(genoPCs)

# Add in chr 1 haplotype cluster covariates
Chr1HapCovMat <- readRDS(file=paste0(project_path, "data/Chr1_hap_incidence_mat", dataset, ".RDS"))

# Subset and order to match individuals in bimbam files, add haplotype cluster covariates
CovMatNamed <- left_join(as.data.frame(sample_names), PreCovMat, by=c("sample_names"="germplasmName")) %>% 
  left_join(Chr1HapCovMat, by=c("sample_names"="germplasmName"))

if(nPCs == 0){CovMatNamed <- CovMatNamed[,-3]}

# Confirm correct sample names and order
sum(sample_names != CovMatNamed$sample_names) == 0

# Write out covariates file 
write.table(CovMatNamed[,-1], 
            file=paste0(project_path, "data/Covariates_", nPCs, "PCs_Chr1Haps_", dataset, ".txt"), 
            col.names=F, row.names=F)
            
# Write out covariates file with names for other analyses
write.table(CovMatNamed, 
            file=paste0(project_path, "data/Covariates_", nPCs, "PCs_Chr1Haps_named_", dataset, ".txt"),
            col.names=T)

```


-----------------------------------------------
_______________________________________________
-----------------------------------------------




## Calculate the relatedness matrix using GEMMA:
  - gk 2 calculates the standardized relationship matrix rather than the centered (gk 1)
  cd /workdir/ssv42/yellowGWAS
```{r}
#for all individuals
system(paste0( 
  "cd ", project_path, "; gemma -g ./data/GenotypeMatrix_", dataset, ".bimbam -p ./data/BLUPmatrix_scaled_", dataset, ".txt -gk 2 -o KMatrix_", dataset))


#check output is correct dimensions
Kmat <-read.table(file=paste0("./output/KMatrix_", dataset, ".sXX.txt"))
dim(Kmat) #378 by 378
dim(Kmat)[1] == length(sample_names)

mean(diag(as.matrix(Kmat))) #1 
range(diag(as.matrix(Kmat))) #0.71 to 1.79
```



# Run associations with TCICHK and DM
-n: specifies which traits to use in the BLUP matrix: TCICHK is column 1 and DM is column 3
-lmm 4: specifies to calculate all the frequentist tests

Note to self: Copy and paste command in the terminal to avoid occupying R while GEMMA is running
```{r, eval=F}
#MV for TC and DM
system(paste0( "cd /workdir/ssv42/mvGWAS; gemma -g ./data/GenotypeMatrix_", dataset, ".bimbam -p ./data/BLUPmatrix_scaled_", dataset, ".txt -k ./output/KMatrix_", dataset,".sXX.txt -a ./data/SNPannotation_", dataset, ".txt  -lmm 4 -n 1 3 -maf 0.05 -o GEMMA_TCDM_", nPCs, "PCs_Scaled_", dataset, "_$(date +'%m%d')"))

#removed covariates file:
#-c ./data/Covariates_", nPCs, "PCs_Chr1Haps_", dataset, ".txt

#Just TC
system(paste0( "cd /workdir/ssv42/mvGWAS; gemma -g ./data/GenotypeMatrix_", dataset, ".bimbam -p ./data/BLUPmatrix_scaled_", dataset, ".txt -k ./output/KMatrix_", dataset,".sXX.txt -a ./data/SNPannotation_", dataset, ".txt -lm 4 -n 1 -maf 0.05 -o GEMMA_TCICHK_", nPCs, "PCs_Scaled_", dataset, "_$(date +'m%d')"))

#Just DM
system(paste0( "cd /workdir/ssv42/mvGWAS; gemma -g ./data/GenotypeMatrix_", dataset, ".bimbam -p ./data/BLUPmatrix_scaled_", dataset, ".txt -k ./output/KMatrix_", dataset,".sXX.txt -a ./data/SNPannotation_", dataset, ".txt -lm 4 -n 3 -maf 0.05 -o GEMMA_DM_", nPCs, "PCs_Scaled_", dataset, "_$(date +'%m%d')"))
```



## Look at the MV results
```{r}
# Get filename of most recent GEMMA run
system(paste0( "ls ", project_path, "output/GEMMA_TCDM_", nPCs, "PCs_Scaled_", dataset, "*.assoc.txt > ./recentruns.txt") )
recentruns <- read.table("./recentruns.txt")
lastrunname <- recentruns[nrow(recentruns), 1]
print(lastrunname)

# Read in results
MVresults <- read.table(lastrunname, header=T) %>%
  mutate(CHR=as.factor(chr), SNP=as.factor(gsub("S","",rs)), BP=ps, P=p_wald)

## QQ plot
qqman::qq(MVresults$P)

## Set cutoffs 
alpha <- 0.05
#Gao modified Bonferroni 
Meff_chrom <- readRDS(file="./data/Meff_perchrom_estimates.RDS")
Meff <- sum(Meff_chrom)
Gao.cutoff <- alpha/Meff


## Prepare the plot
MVresults_sort <- MVresults %>% arrange(chr, ps)
#order factor levels so they display in order of chromosome
MVresults$SNP <- factor(MVresults$SNP, levels = MVresults$SNP)

#set color scheme
colPalette <- rep(c("blue4","pink3"),9)
#set positions for axis ticks centered by chromosome
xaxisticks <- MVresults %>% group_by(CHR) %>% summarise(midpt = nth(SNP, round(n()/2))) %>% as.data.frame() 


## Plot Manhattan with raw p-wald values
ggplot(MVresults, aes(x=SNP, y=-log10(p_wald))) +
  geom_point(aes(color=CHR), size=0.5) +
  scale_color_manual(values=colPalette) +
  labs(y="-log10(p)") +
  scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
  ggtitle(paste0("GEMMA mvGWAS for TC and DM, ", nPCs, "PCs, ", dataset, " dataset")) +
  guides(color="none")+ 
  geom_vline(xintercept="1_24155945", color="orange", lty="dashed") + #closest SNP to PSY2
  geom_hline(yintercept = -log10(Gao.cutoff), linetype="dashed", lwd=0.4)

ggsave(filename=paste0("./figures/MVmanhattan_", nPCs, "PCs_Scaled_", dataset, "_", Sys.Date(),".jpg"))



## Examine distribution of beta estimate signs 
#subset for SNPs where the marker estimates have opposite signs
MVresults_betas <- MVresults %>% mutate(beta_sign = 
                        case_when(beta_1 > 0 & beta_2 < 0 | 
                                    beta_1 < 0 & beta_2 > 0 ~ "opposite",
                                  beta_1 >= 0 & beta_2 >= 0 | 
                                    beta_1 <= 0 & beta_2 <= 0 ~ "same or none")) %>%
  mutate(sig = case_when( P <= Gao.cutoff ~ "significant", TRUE ~ "n.s."))

 
ggplot(MVresults_betas, aes(x=SNP, y=-log10(p_wald))) +
  geom_point(aes(color=beta_sign, pch=sig)) +
  scale_shape_manual(values = c(19, 8)) +
  labs(y="-log10(p))", color="effect direction", shape="significance") +
  scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
  ggtitle(paste0("Effect directions for GEMMA mvGWAS for TC and DM ", nPCs, "PCs, ", dataset, " dataset")) +
  geom_vline(xintercept="1_24154529", color="orange", lty="dashed") #closest to PSY2

ggsave(filename=paste0("./figures/MVmanhattan_EffectDir_", nPCs, "PCs_Scaled_", dataset, "_", Sys.Date(),".jpg"))



# Look at Z-scores and p-values
MVresults$Z_TC <- MVresults$beta_1/sqrt(MVresults$Vbeta_1_1)

MVresults$Z_DM <- MVresults$beta_2/sqrt(MVresults$Vbeta_2_2)


ggplot(MVresults, aes(Z_TC, Z_DM, color=-log10(P))) +
  geom_point() +
  ggtitle("GEMMA p-value")

MVresults$Zdist <- sqrt(MVresults$Z_TC^2 + MVresults$Z_DM^2)

MVresults$Zdist_p <- 2*pnorm(MVresults$Zdist, mean=0, sd=1, lower.tail=F)


ggplot(MVresults, aes(Z_TC, Z_DM, color=-log10(Zdist_p))) +
  geom_point() +
  ggtitle("Z-distance p-value")

ggplot(MVresults, aes(P, Zdist_p)) +
  geom_point(aes(color=Z_DM)) +
  xlab("GEMMA p-value") +  
  ylab("p-value from bivar. Z-score distance from origin")


#Multivariate null distribution: with beta correlation
betacor <- cor(MVresults$beta_1, MVresults$beta_2) #is this right to include the correlation in the null dist?
cov_matrix <- matrix(c(1, betacor, betacor, 1), nrow=2, ncol=2)
nullZs <- as.data.frame(mvrnorm(n=500000, mu=c(0,0), Sigma=cov_matrix))
colnames(nullZs) <- c("Z1", "Z2")
nullZs$Zdist <- sqrt(nullZs$Z1^2 + nullZs$Z2^2)

ggplot(nullZs, aes(Z1, Z2, color=Zdist)) +
  geom_point() +
  ggtitle("null z-scores")

#0.05 = 25000/500000

alpha0.05cutoff <- nullZs %>% arrange(-Zdist) %>% head(n=25000) %>% dplyr::select(Zdist) %>% min(.)

hist(nullZs$Zdist, main="Null distribution of bivariate Z-score distance w/ beta correlation", xlab="bivariate Z-score distance from origin")
abline(v= alpha0.05cutoff, lty=3)




#Multivariate null distribution: without beta correlation
cov_matrix <- matrix(c(1, 0, 0, 1), nrow=2, ncol=2)
nullZs <- as.data.frame(mvrnorm(n=500000, mu=c(0,0), Sigma=cov_matrix))
colnames(nullZs) <- c("Z1", "Z2")
nullZs$Zdist <- sqrt(nullZs$Z1^2 + nullZs$Z2^2)

ggplot(nullZs, aes(Z1, Z2, color=Zdist)) +
  geom_point() +
  ggtitle("null z-scores")

alpha0.05cutoff <- nullZs %>% arrange(-Zdist) %>% head(n=25000) %>% dplyr::select(Zdist) %>% min(.)

hist(nullZs$Zdist, main="Null distribution of bivariate Z-score distance without beta correlation", xlab="bivariate Z-score distance from origin")
abline(v= alpha0.05cutoff, lty=3)



```

Compare with asreml model:
```{r}
setwd("/workdir/ssv42/mvGWAS/")
project_path <- getwd()
dataset <- "CASS"



  library(tidyverse); library(asreml)
  project_path <<- "/workdir/ssv42/mvGWAS/"
  dataset <<- "CASS"
  
  #read in data
  genomat_ROG <<- readRDS(file=paste0(project_path, "data/genomat_restofgenome_", dataset, ".RDS"))
  Amatrix_notsing <<- readRDS(file=paste0(project_path, "data/Amatrix_notsingular_", dataset, ".RDS"))
  BLUPs_hap_data <<- readRDS(file=paste0(project_path, "data/BLUPs_scaled_and_hap_data_", dataset, ".RDS"))
  # Amat_haps_PSY2_notsing <<- readRDS(file=paste0(project_path, "data/Amat_PSY2haps_notsingular_", dataset, ".RDS"))
  # Amat_haps_IDMs_notsing <<- readRDS(file=paste0(project_path, "data/Amat_IDMhaps_notsingular_", dataset, ".RDS"))

  # Loop across all "rest of genome" SNPs 
  #storage for SNP effects
  SNPeff_storage <- list()
  #storage for Wald test of fixed effect significance
  WaldTest_storage <- list()

  for(i in seq_along(genomat_ROG$SNP)){
    
    SNPi <<- genomat_ROG$SNP[i]
    SNPidata <- as.data.frame(t(genomat_ROG[i, -c(1:3)]))
    colnames(SNPidata) <- paste0(SNPi)
    SNPidata$germplasmName <- gsub("\\.","-", rownames(SNPidata))

    modeldata <- left_join(BLUPs_hap_data, SNPidata, by="germplasmName") %>% 
      mutate(germplasmName = as.factor(germplasmName))
    
    #model calculating SNP effect on bivariate trait
    SNPimodel <- asreml(data = modeldata, 
        formula(paste0("cbind(TCICHK, DM) ~ trait + ", SNPi)), 
         random = ~ vm(germplasmName, Amatrix_notsing) , maxit=20)

    
    WaldTest <- wald.asreml(SNPimodel, ssType = "conditional", denDF = "numeric")
    
    WaldTest_storage[[i]] <- WaldTest$Wald
    
    SNPeff_storage[[i]] <- summary(SNPimodel, coef = TRUE)$coef.fixed
    
    
  saveRDS(SNPeff_storage, file=paste0(project_path, 
                  "output/mvGWAS_asreml_output/asreml_simple_SNPeff_", 
              format(Sys.time(), "%m%d"), ".RDS"))
  saveRDS(WaldTest_storage, file=paste0(project_path, 
                  "output/mvGWAS_asreml_output/asreml_simple_SNPWaldTest_", 
              format(Sys.time(), "%m%d"), ".RDS"))

  cat(paste0(SNPi, ", ", i, " done \n"),
      file = paste0(project_path, 
                    "output/mvGWAS_asreml_output/asreml_simple_progressbar.txt"), append=T)

  } #end SNP





```



# Look at the univariate results for TC
```{r}
# Get filename of most recent GEMMA run
system(paste0( "ls ", project_path, "output/GEMMA_TCICHK_", nPCs, "PCs_Scaled_", dataset, "*.assoc.txt > ./recentruns.txt") )
recentruns <- read.table("./recentruns.txt")
lastrunname <- recentruns[nrow(recentruns), 1] #need to change this selection for the last run
print(lastrunname)

# Read in results
TCresults <- read.table(lastrunname, header=T) %>%
  mutate(CHR=as.factor(chr), SNP=as.factor(gsub("S","",rs)), BP=ps, P=p_wald)

## QQ plot
#png(filename=paste0(project_path, "figures/QQ-plot_TCICHK_", nPCs, "PCs_Chr1Haps_", dataset, ".png"))
  qqman::qq(TCresults$P, main=paste0("QQ-plot for TCICHK, ", nPCs, "PCs and no Chr1 Haps, ", dataset, " data"))
#dev.off()

## Prepare the plot
#order factor levels so they display in order of chromosome
TCresults$SNP <- factor(TCresults$SNP, levels = TCresults$SNP)
#set color scheme
colPalette <- rep(c("blue4","pink3"),9)
#set positions for axis ticks centered by chromosome
xaxisticks <- TCresults %>% group_by(CHR) %>% summarise(midpt = nth(SNP, round(n()/2))) %>% as.data.frame() 

## Set cutoffs 
alpha <- 0.05
#Gao modified Bonferroni 
Meff_chrom <- readRDS(file="./data/Meff_perchrom_estimates.RDS")
Meff <- sum(Meff_chrom)
Gao.cutoff <- alpha/Meff


## Plot Manhattan
ggplot(TCresults, aes(x=SNP, y=-log10(p_wald))) +
  geom_point(aes(color=CHR), size=0.5) +
  scale_color_manual(values=colPalette) +
  labs(y="-log10(p)") +
  scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
  ggtitle(paste0("Univariate GWAS for TC with ", nPCs, " PCs and Chr1 Haps, ", dataset," dataset")) +
  guides(color="none") +
  geom_hline(yintercept = -log10(Gao.cutoff), linetype="dashed", lwd=0.4)+ 
  geom_vline(xintercept="1_24155945", color="orange", lty="dashed")

#ggsave(filename=paste0("./figures/UVmanhattan_TCICHK_", nPCs, "PCs_Chr1Haps_", dataset, "_", Sys.Date(),".jpg"))

```



# Look at the univariate results for DM
```{r}
# Get filename of most recent GEMMA run
system(paste0( "ls ", project_path, "output/GEMMA_DM_", nPCs, "PCs_Chr1Haps_", dataset, "*.assoc.txt > ./recentruns.txt") )
recentruns <- read.table("./recentruns.txt")
lastrunname <- recentruns[nrow(recentruns), 1] #need to change this selection for the last run
print(lastrunname)

# Read in results
DMresults <- read.table(lastrunname, header=T) %>%
  mutate(CHR=as.factor(chr), SNP=as.factor(gsub("S","",rs)), BP=ps, P=p_wald)

## QQ plot
png(filename=paste0(project_path, "figures/QQ-plot_DM_", nPCs, "PCs_Chr1Haps_", dataset, ".png"))
  qqman::qq(DMresults$P, main=paste0("QQ-plot for DM, ", nPCs, "PCs and Chr1 Haps, ", dataset, " data"))
dev.off()

## Prepare the plot
#order factor levels so they display in order of chromosome
DMresults$SNP <- factor(DMresults$SNP, levels = DMresults$SNP)
#set color scheme
colPalette <- rep(c("blue4","pink3"),9)
#set positions for axis ticks centered by chromosome
xaxisticks <- DMresults %>% group_by(CHR) %>% summarise(midpt = nth(SNP, round(n()/2))) %>% as.data.frame() 


## Plot Manhattan
ggplot(DMresults, aes(x=SNP, y=-log10(p_wald))) +
  geom_point(aes(color=CHR), size=0.5) +
  scale_color_manual(values=colPalette) +
  labs(y="-log10(p)") +
  scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
  ggtitle(paste0("Univariate GWAS for DM with ", nPCs, " PCs and Chr1 Haps, ", dataset," data")) +
  guides(color="none") +
  geom_hline(yintercept = -log10(Gao.cutoff), linetype="dashed", lwd=0.4)+ 
  geom_vline(xintercept="1_24155945", color="orange", lty="dashed")

ggsave(filename=paste0("./figures/UVmanhattan_DM_", nPCs, "PCs_Chr1Haps_", dataset, "_", Sys.Date(),".jpg"))

```

Merge GEMMA multivariate and univariate results
```{r}
GEMMA_allresults <- left_join(MVresults, TCresults, by=c("chr","rs","ps","allele1","allele0","CHR","SNP","BP")) %>%
  left_join(DMresults, by=c("chr","rs","ps","allele1","allele0","CHR","SNP","BP")) %>%
  rename("mvP" = "P.x", "tcP" = "P.y", "dmP" = "P") %>%
  rename("beta_tc" = "beta.x", "se_tc" = "se.x", "beta_dm" = "beta.y", "se_dm" = "se.y") %>%
  rename("mvP" = "p_wald.x")
  
saveRDS(GEMMA_allresults, file=paste0(project_path, "output/GEMMA_combined_", nPCs, "PCs_", dataset, "_", format(Sys.time(), "%Y%m%d_%H%M"), ".RDS"))
```



Set top Chr 1 SNP as a covariate:
```{r}
# Read in PC decomposition
genoPCA <- readRDS(file=paste0(project_path, "data/DosageMatrix_PCA_", dataset, ".RDS")) 
genoPCs <- as.data.frame(genoPCA$x)

# Make covariates file with first 3 PCs
nPCs <- 3
PreCovMat <- as.data.frame(cbind(rep(1,dim(genoPCs)[1]), genoPCs[,1:nPCs]))
colnames(PreCovMat)[1] <- "intercept"
PreCovMat$germplasmName <- rownames(genoPCs)

# Get top chr 1 SNP genotypes
#top SNP in combined gemma/Stephens: "S1_26700297"
#top SNP near PSY2 in univariate GEMMA for DM: "S1_23985098"
bimbam <- read.table(file=paste0(project_path, "data/GenotypeMatrix_Named_", dataset, ".bimbam"), header=T)
topchr1SNP <- bimbam[which(bimbam$SNP == "S1_23985098"), -c(1:3)] %>% t() %>% as.data.frame()
topchr1SNP$germplasmName <- gsub("\\.","-", rownames(topchr1SNP))

# Subset and order to match individuals in bimbam files
CovMatNamed <- left_join(as.data.frame(sample_names), PreCovMat, by=c("sample_names"="germplasmName")) %>% 
  left_join(topchr1SNP,  by=c("sample_names"="germplasmName"))

# Confirm correct sample names and order
sum(sample_names != CovMatNamed$sample_names) == 0

# Write out covariates file 
write.table(CovMatNamed[,-1], 
            file=paste0(project_path, "data/Covariates_", nPCs, "PCs_S1-23985098_", dataset, ".txt"), 
            col.names=F, row.names=F)
            
# Write out covariates file with names for other analyses
write.table(CovMatNamed, 
            file=paste0(project_path, "data/Covariates_", nPCs, "PCs_S1-23985098_", dataset, ".txt"), 
            col.names=T)

```


Re-run GEMMA with chr 1 top SNP as covariate:
```{r}

#MV for TC and DM
system(paste0( "cd /workdir/ssv42/mvGWAS; gemma -g ./data/GenotypeMatrix_", dataset, ".bimbam -p ./data/BLUPmatrix_", dataset, ".txt -k ./output/KMatrix_", dataset,".sXX.txt -a ./data/SNPannotation_", dataset, ".txt -c ./data/Covariates_", nPCs, "PCs_S1-23985098_", dataset, ".txt -lmm 4 -n 1 3 -maf 0.05 -o GEMMA_TCDM_", nPCs, "PCs_S1-23985098_", dataset, "_$(date +'%Y%m%d_%H%M')"))


#Just TC
system(paste0( "cd /workdir/ssv42/mvGWAS; gemma -g ./data/GenotypeMatrix_", dataset, ".bimbam -p ./data/BLUPmatrix_", dataset, ".txt -k ./output/KMatrix_", dataset,".sXX.txt -a ./data/SNPannotation_", dataset, ".txt -c ./data/Covariates_", nPCs, "PCs_S1-23985098_", dataset, ".txt -lm 4 -n 1 -maf 0.05 -o GEMMA_TCICHK_", nPCs, "PCs_S1-23985098_", dataset, "_$(date +'%Y%m%d_%H%M')"))

#Just DM
system(paste0( "cd /workdir/ssv42/mvGWAS; gemma -g ./data/GenotypeMatrix_", dataset, ".bimbam -p ./data/BLUPmatrix_", dataset, ".txt -k ./output/KMatrix_", dataset,".sXX.txt -a ./data/SNPannotation_", dataset, ".txt -c ./data/Covariates_", nPCs, "PCs_S1-23985098_", dataset, ".txt -lm 4 -n 3 -maf 0.05 -o GEMMA_DM_", nPCs, "PCs_S1-23985098_", dataset, "_$(date +'%Y%m%d_%H%M')"))
```

# Look at the MV results with chr. 1 covariate
```{r}
# Get filename of most recent GEMMA run
system(paste0( "ls ", project_path, "output/GEMMA_TCDM_", nPCs, "PCs_S1-26700297_", dataset, "*.assoc.txt > ./recentruns.txt") )
recentruns <- read.table("./recentruns.txt")
lastrunname <- recentruns[nrow(recentruns), 1] #need to change this selection for the last run
print(lastrunname)

# Read in results
MVresults <- read.table(lastrunname, header=T) %>%
  mutate(CHR=as.factor(chr), SNP=as.factor(gsub("S","",rs)), BP=ps, P=p_wald)

## QQ plot
qqman::qq(MVresults$P)

## Set cutoffs 
alpha <- 0.05
#Gao modified Bonferroni 
Meff_chrom <- readRDS(file="./data/Meff_perchrom_estimates.RDS")
Meff <- sum(Meff_chrom)
Gao.cutoff <- alpha/Meff


## Prepare the plot
MVresults_sort <- MVresults %>% arrange(chr, ps)
#order factor levels so they display in order of chromosome
MVresults$SNP <- factor(MVresults$SNP, levels = MVresults$SNP)

#set color scheme
colPalette <- rep(c("blue4","pink3"),9)
#set positions for axis ticks centered by chromosome
xaxisticks <- MVresults %>% group_by(CHR) %>% summarise(midpt = nth(SNP, round(n()/2))) %>% as.data.frame() 


## Plot Manhattan with raw p-wald values
ggplot(MVresults, aes(x=SNP, y=-log10(p_wald))) +
  geom_point(aes(color=CHR), size=0.5) +
  scale_color_manual(values=colPalette) +
  labs(y="-log10(p)") +
  scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
  ggtitle(paste0("GEMMA mvGWAS for TC and DM, ", nPCs, "PCs, S1-26700297 covariate, ", dataset, " dataset")) +
  guides(color="none")+ 
  geom_vline(xintercept="1_24155945", color="orange", lty="dashed") + #closest SNP to PSY2
  geom_hline(yintercept = -log10(Gao.cutoff), linetype="dashed", lwd=0.4)

ggsave(filename=paste0("./figures/MVmanhattan_", nPCs, "PCs_S1-26700297_", dataset, "_", Sys.Date(),".jpg"))



## Examine distribution of beta estimate signs 
#subset for SNPs where the marker estimates have opposite signs
MVresults_betas <- MVresults %>% mutate(beta_sign = 
                        case_when(beta_1 > 0 & beta_2 < 0 | 
                                    beta_1 < 0 & beta_2 > 0 ~ "opposite",
                                  beta_1 >= 0 & beta_2 >= 0 | 
                                    beta_1 <= 0 & beta_2 <= 0 ~ "same or none")) %>%
  mutate(sig = case_when( P <= Gao.cutoff ~ "significant", TRUE ~ "n.s."))

 
ggplot(MVresults_betas, aes(x=SNP, y=-log10(p_wald))) +
  geom_point(aes(color=beta_sign, pch=sig)) +
  scale_shape_manual(values = c(19, 8)) +
  labs(y="-log10(p))", color="effect direction", shape="significance") +
  scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
  ggtitle(paste0("Effect directions for GEMMA mvGWAS for TC and DM ", nPCs, "PCs, ", dataset, " dataset")) +
  geom_vline(xintercept="1_24154529", color="orange", lty="dashed") #closest to PSY2

ggsave(filename=paste0("./figures/MVmanhattan_EffectDir_", nPCs, "PCs_", dataset, "_", Sys.Date(),".jpg"))
```






# Look at the univariate results for TC
```{r}
# Get filename of most recent GEMMA run
system(paste0( "ls ", project_path, "output/GEMMA_TCICHK_", nPCs, "PCs_S1-26700297_", dataset, "*.assoc.txt > ./recentruns.txt") )
recentruns <- read.table("./recentruns.txt")
lastrunname <- recentruns[nrow(recentruns), 1] 
print(lastrunname)

# Read in results
TCresults <- read.table(lastrunname, header=T) %>%
  mutate(CHR=as.factor(chr), SNP=as.factor(gsub("S","",rs)), BP=ps, P=p_wald)

## QQ plot
qqman::qq(TCresults$P)


## Prepare the plot
#order factor levels so they display in order of chromosome
TCresults$SNP <- factor(TCresults$SNP, levels = TCresults$SNP)
#set color scheme
colPalette <- rep(c("blue4","pink3"),9)
#set positions for axis ticks centered by chromosome
xaxisticks <- TCresults %>% group_by(CHR) %>% summarise(midpt = nth(SNP, round(n()/2))) %>% as.data.frame() 


## Plot Manhattan
ggplot(TCresults, aes(x=SNP, y=-log10(p_wald))) +
  geom_point(aes(color=CHR), size=0.5) +
  scale_color_manual(values=colPalette) +
  labs(y="-log10(p)") +
  scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
  ggtitle(paste0("Univariate GWAS for TC with ", nPCs, " PCs and S1-26700297 covariate, ", dataset," dataset")) +
  guides(color="none") +
  geom_hline(yintercept = -log10(Gao.cutoff), linetype="dashed", lwd=0.4)+ 
  geom_vline(xintercept="1_24155945", color="orange", lty="dashed")

ggsave(filename=paste0("./figures/UVmanhattan_TCICHK_", nPCs, "PCs_S1-26700297_", dataset, "_", Sys.Date(),".jpg"))

```

# Look at the univariate results for DM
```{r}
# Get filename of most recent GEMMA run
system(paste0( "ls ", project_path, "output/GEMMA_DM_", nPCs, "PCs_S1-26700297_", dataset, "*.assoc.txt > ./recentruns.txt") )
recentruns <- read.table("./recentruns.txt")
lastrunname <- recentruns[nrow(recentruns), 1] #need to change this selection for the last run
print(lastrunname)

# Read in results
DMresults <- read.table(lastrunname, header=T) %>%
  mutate(CHR=as.factor(chr), SNP=as.factor(gsub("S","",rs)), BP=ps, P=p_wald)

## QQ plot
qqman::qq(DMresults$P)


## Prepare the plot
#order factor levels so they display in order of chromosome
DMresults$SNP <- factor(DMresults$SNP, levels = DMresults$SNP)
#set color scheme
colPalette <- rep(c("blue4","pink3"),9)
#set positions for axis ticks centered by chromosome
xaxisticks <- DMresults %>% group_by(CHR) %>% summarise(midpt = nth(SNP, round(n()/2))) %>% as.data.frame() 


## Plot Manhattan
ggplot(DMresults, aes(x=SNP, y=-log10(p_wald))) +
  geom_point(aes(color=CHR), size=0.5) +
  scale_color_manual(values=colPalette) +
  labs(y="-log10(p)") +
  scale_x_discrete(name="chromosome", breaks=xaxisticks$midpt, labels=c(1:18)) +
  ggtitle(paste0("Univariate GWAS for DM with ", nPCs, " PCs and S1-26700297 covariate, ", dataset," data")) +
  guides(color="none") +
  geom_hline(yintercept = -log10(Gao.cutoff), linetype="dashed", lwd=0.4)+ 
  geom_vline(xintercept="1_24155945", color="orange", lty="dashed")

ggsave(filename=paste0("./figures/UVmanhattan_DM_", nPCs, "PCs_", dataset, "_", Sys.Date(),".jpg"))

```










How I calculated Meff and modified Bonferonni cutoff a la Gao et al 2008:
```{r, eval=F}
#pairwise SNP correlation matrix: 66k x 66k matrix -- cor()
#PCA eigenvalue decomposition -- eigen()
#select eigenvalues that explain  99.5%  of  the  variation  for  SNP  data
#the number of selected eigenvalues = M_eff
#adjusted bonferroni cutoff = alpha/M_eff

#read in genotype matrix, keep coding as {0, 1, 2}

genomat_t <- read.table("/workdir/ssv42/mvGWAS/data/GenotypeMatrix_Named_CASS.bimbam") %>%
  dplyr::select(-c(1:3)) %>% as.matrix() %>% t(.)

#count how many SNPs per chromosome
SNPorder <- read.table("/workdir/ssv42/mvGWAS/data/SNPorder_CASS.txtFALSEFALSE")
SNPsPerChr <- c()
for (i in c(1:18)){
  nSNPs <- sum(grepl(paste0("S",i,"_"), SNPorder[,1]))
  SNPsPerChr <- c(SNPsPerChr, nSNPs)
}

#Subset SNPs per chromosome
Meff_chrom <- c()
Meff_chrom_poolr <- c()
start <- 1
for (i in seq_along(SNPsPerChr)){
  print(i)
  
  end <- start + SNPsPerChr[i] - 1
  
  genomat_sub <- genomat_t[,c(start:end)]
  
  PCAblock <- summary(prcomp(genomat_sub))$importance
  Meff_i <- min(which(PCAblock["Cumulative Proportion",] >= 0.9950))
  
  geno_cor_i <- cor(genomat_sub)
  Meff_poolr_i <- poolr::meff(geno_cor_i, method="gao")
  
  Meff_chrom <- c(Meff_chrom, Meff_i)
  Meff_chrom_poolr <- c(Meff_chrom_poolr, Meff_poolr_i)
  
  start <- end + 1
}

saveRDS(Meff_chrom, file="/workdir/ssv42/mvGWAS/output/Meff_perchrom_estimates_CASS.RDS")

Meff_total <- sum(Meff_chrom)
print(Meff_total) #4661

#previously: 4575
SNPcormat <- cor(genomat_t[,1:10000])
#saveRDS(SNPcormat, file="/workdir/ssv42/mvGWAS/output/SNPcorrelationmatrix.RDS")

#alternative method:
poolr::meff(SNPcormat, method = "gao")

```
